File: VB/InfiniteAsyncSourceMVVMSample/IssueService/IssuesService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks

Namespace InfiniteAsyncSourceMVVMSample

    Public Module IssuesService

#Region "helpers"
        Private SyncObject As Object = New Object()

        Private AllIssues As Lazy(Of List(Of IssueData)) = New Lazy(Of List(Of IssueData))(Function()
            Dim [date] = Date.Today
            Dim rnd = New Random(0)
            Return Enumerable.Range(0, 100000).[Select](Function(i)
                [date] = [date].AddSeconds(-rnd.Next(20 * 60))
                Return New IssueData(id:=i, subject:=GetSubject(), userId:=GetFromId(), created:=[date], votes:=rnd.Next(100), priority:=GetPriority())
            End Function).ToList()
        End Function)

        Private Class DefaultSortComparer
            Implements IComparer(Of IssueData)

            Private Function Compare(ByVal x As IssueData, ByVal y As IssueData) As Integer Implements IComparer(Of IssueData).Compare
                If x.Created.Date <> y.Created.Date Then Return Comparer(Of Date).Default.Compare(x.Created.Date, y.Created.Date)
                Return Comparer(Of Integer).Default.Compare(x.Votes, y.Votes)
            End Function
        End Class

#End Region
        Public Async Function GetIssuesAsync(ByVal skip As Integer, ByVal take As Integer, ByVal sortOrder As IssueSortOrder, ByVal filter As IssueFilter) As Task(Of IssueData())
            Await Task.Delay(300).ConfigureAwait(False)
            Dim issues = SortIssues(sortOrder, AllIssues.Value)
            If filter IsNot Nothing Then issues = FilterIssues(filter, issues)
            Return issues.Skip(skip).Take(take).[Select](Function(x) x.Clone()).ToArray()
        End Function

        Public Async Function GetUsersAsync() As Task(Of UserData())
            Await Task.Delay(300).ConfigureAwait(False)
            Return Users.[Select](Function(x, i)
                Dim split = x.Split(" "c)
                Return New UserData(i, split(0), split(1))
            End Function).ToArray()
        End Function

        Public Async Function GetSummariesAsync(ByVal filter As IssueFilter) As Task(Of IssuesSummaries)
            Await Task.Delay(300).ConfigureAwait(False)
            Dim issues = CType(AllIssues.Value, IEnumerable(Of IssueData))
            If filter IsNot Nothing Then issues = FilterIssues(filter, issues)
            Dim lastCreated = If(issues.Any(), issues.Max(Function(x) x.Created), DirectCast(Nothing, Date?))
            Return New IssuesSummaries(count:=issues.Count(), lastCreated:=lastCreated)
        End Function

        Public Async Function UpdateRowAsync(ByVal row As IssueData) As Task
            If row Is Nothing Then Return
            Dim data As IssueData = AllIssues.Value.FirstOrDefault(Function(x) x.Id = row.Id)
            If data Is Nothing Then Return
            data.Priority = row.Priority
            data.Subject = row.Subject
            data.UserId = row.UserId
            data.Votes = row.Votes
            data.Created = row.Created
            Await Task.Delay(500).ConfigureAwait(False)
        End Function

        Public Async Function AddNewRowAsync(ByVal row As IssueData) As Task
            Await Task.Delay(500).ConfigureAwait(False)
            AllIssues.Value.Insert(0, row)
            row.Id = AllIssues.Value.Max(Function(x) x.Id) + 1
        End Function

        Public Async Function DeleteRowAsync(ByVal row As IssueData) As Task
            Await Task.Delay(500).ConfigureAwait(False)
            AllIssues.Value.Remove(AllIssues.Value.Find(Function(x) x.Id = row.Id))
        End Function

#Region "filter"
        Private Function FilterIssues(ByVal filter As IssueFilter, ByVal issues As IEnumerable(Of IssueData)) As IEnumerable(Of IssueData)
            If filter.CreatedFrom IsNot Nothing OrElse filter.CreatedTo IsNot Nothing Then
                If filter.CreatedFrom Is Nothing OrElse filter.CreatedTo Is Nothing Then
                    Throw New InvalidOperationException()
                End If

                issues = issues.Where(Function(x) x.Created >= filter.CreatedFrom.Value AndAlso x.Created < filter.CreatedTo)
            End If

            If filter.MinVotes IsNot Nothing Then
                issues = issues.Where(Function(x) x.Votes >= filter.MinVotes.Value)
            End If

            If filter.Priority IsNot Nothing Then
                issues = issues.Where(Function(x) x.Priority = filter.Priority)
            End If

            Return issues
        End Function

#End Region
#Region "sort"
        Private Function SortIssues(ByVal sortOrder As IssueSortOrder, ByVal issues As IEnumerable(Of IssueData)) As IEnumerable(Of IssueData)
            Select Case sortOrder
                Case IssueSortOrder.Default
                    Return issues '.OrderByDescending(x => x, new DefaultSortComparer()).ThenByDescending(x => x.Created);
                Case IssueSortOrder.CreatedDescending
                    Return issues.OrderByDescending(Function(x) x.Created)
                Case IssueSortOrder.VotesAscending
                    Return issues.OrderBy(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case IssueSortOrder.VotesDescending
                    Return issues.OrderByDescending(Function(x) x.Votes).ThenByDescending(Function(x) x.Created)
                Case Else
                    Throw New InvalidOperationException()
            End Select
        End Function
#End Region
    End Module
End Namespace

File: CS/InfiniteAsyncSourceMVVMSample/IssueService/IssuesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace InfiniteAsyncSourceMVVMSample {
    public static class IssuesService {
        #region helpers
        static object SyncObject = new object();
        static Lazy<IssueData[]> AllIssues = new Lazy<IssueData[]>(() => {
            var date = DateTime.Today;
            var rnd = new Random(0);
            return Enumerable.Range(0, 100000)
                .Select(i => {
                    date = date.AddSeconds(-rnd.Next(20 * 60));
                    return new IssueData(
                        id: i,
                        subject: OutlookDataGenerator.GetSubject(),
                        userId: OutlookDataGenerator.GetFromId(),
                        created: date,
                        votes: rnd.Next(100),
                        priority: OutlookDataGenerator.GetPriority());
                }).ToArray();
        });
        class DefaultSortComparer : IComparer<IssueData> {
            int IComparer<IssueData>.Compare(IssueData x, IssueData y) {
                if(x.Created.Date != y.Created.Date)
                    return Comparer<DateTime>.Default.Compare(x.Created.Date, y.Created.Date);
                return Comparer<int>.Default.Compare(x.Votes, y.Votes);
            }
        } 
        #endregion

        public async static Task<IssueData[]> GetIssuesAsync(int skip, int take, IssueSortOrder sortOrder, IssueFilter filter) {
            await Task.Delay(300).ConfigureAwait(false);
            var issues = SortIssues(sortOrder, AllIssues.Value);
            if(filter != null)
                issues = FilterIssues(filter, issues);
            return issues.Skip(skip).Take(take).Select(x => x.Clone()).ToArray();
        }

        public async static Task<UserData[]> GetUsersAsync() {
            await Task.Delay(300).ConfigureAwait(false);
            return OutlookDataGenerator.Users
                .Select((x, i) => {
                    var split = x.Split(' ');
                    return new UserData(i, split[0], split[1]);
                })
                .ToArray();
        }

        public async static Task<IssuesSummaries> GetSummariesAsync(IssueFilter filter) {
            await Task.Delay(300).ConfigureAwait(false);
            var issues = (IEnumerable<IssueData>)AllIssues.Value;
            if(filter != null)
                issues = FilterIssues(filter, issues);
            var lastCreated = issues.Any() ? issues.Max(x => x.Created) : default(DateTime?);
            return new IssuesSummaries(count: issues.Count(), lastCreated: lastCreated);
        }

        public async static Task UpdateRowAsync(IssueData row) {
            if(row == null)
                return;
            IssueData data = AllIssues.Value.FirstOrDefault(x => x.Id == row.Id);
            if(data == null)
                return;
            data.Priority = row.Priority;
            data.Subject = row.Subject;
            data.UserId = row.UserId;
            data.Votes = row.Votes;
            data.Created = row.Created;
            await System.Threading.Tasks.Task.Delay(500).ConfigureAwait(false);
        }

        #region filter
        static IEnumerable<IssueData> FilterIssues(IssueFilter filter, IEnumerable<IssueData> issues) {
            if(filter.CreatedFrom != null || filter.CreatedTo != null) {
                if(filter.CreatedFrom == null || filter.CreatedTo == null) {
                    throw new InvalidOperationException();
                }
                issues = issues.Where(x => x.Created >= filter.CreatedFrom.Value && x.Created < filter.CreatedTo);
            }
            if(filter.MinVotes != null) {
                issues = issues.Where(x => x.Votes >= filter.MinVotes.Value);
            }
            if(filter.Priority != null) {
                issues = issues.Where(x => x.Priority == filter.Priority);
            }
            return issues;
        }
        #endregion

        #region sort
        static IEnumerable<IssueData> SortIssues(IssueSortOrder sortOrder, IEnumerable<IssueData> issues) {
            switch(sortOrder) {
            case IssueSortOrder.Default:
                return issues;//.OrderByDescending(x => x, new DefaultSortComparer()).ThenByDescending(x => x.Created);
            case IssueSortOrder.CreatedDescending:
                return issues.OrderByDescending(x => x.Created);
            case IssueSortOrder.VotesAscending:
                return issues.OrderBy(x => x.Votes).ThenByDescending(x => x.Created);
            case IssueSortOrder.VotesDescending:
                return issues.OrderByDescending(x => x.Votes).ThenByDescending(x => x.Created);
            default:
                throw new InvalidOperationException();
            }
        } 
        #endregion
    }
}
